tests/test_mesos.py

```diff
@@ -0,0 +1,59 @@
+# stdlib imports
+import re
+
+# third-party imports
+import responses
+
+# local imports
+from shpkpr.mesos import resolve_leader_url
+from shpkpr.mesos import MesosClient
+
+
+def _mock_valid_redirect():
+    responses.add(responses.HEAD,
+                  'http://master.example.com:5050/master/redirect',
+                  status=307,
+                  adding_headers={'Location': '//leader.example.com:5050'})
+    responses.add(responses.HEAD,
+                  re.compile(r'http://leader\.example\.com:5050/?'),
+                  status=200)
+
+
+@responses.activate
+def test_resolve_leader_valid_redirect():
+    _mock_valid_redirect()
+
+    leader_url = resolve_leader_url('http://master.example.com:5050')
+    assert leader_url == "http://leader.example.com:5050"
+
+
+@responses.activate
+def test_resolve_leader_valid_redirect_with_trailing_slash():
+    _mock_valid_redirect()
+
+    leader_url = resolve_leader_url('http://master.example.com:5050/')
+    assert leader_url == "http://leader.example.com:5050"
+
+
+@responses.activate
+def test_mesos_client_resolves_leader_url():
+    _mock_valid_redirect()
+
+    client = MesosClient('http://master.example.com:5050')
+    # testing private members of a class isn't usually encouraged, however, in
+    # this case the private members under test are the parts we've overridden
+    # from dcos.mesos.DCOSClient and we need to test our implementation.
+    assert client._leader_url is None
+    assert client._mesos_master_url == "http://leader.example.com:5050"
+    assert client._leader_url == "http://leader.example.com:5050"
```

Comment:
> Seems like we're asserting that the call to `_mesos_master_url` changes the property `_leader_url`

So that's exactly what we're doing here.

`_mesos_master_url` is an internal property used by `dcos.mesos.DCOSClient`. It's set in the `__init__` method of that class and loaded directly from a configuration file. The class uses this attribute in mutliple places. In our `MesosClient` class we've implemented `_mesos_master_url` as a dynamic read-only property that automatically resolves to the current leader url and uses the `_leader_url` property to cache the result (so it only has to do a single HTTP lookup).

The more I work with the `dcos` library the more I realise that relying on it long-term isn't a viable proposition, we're currently relying on overriding and patching too many internal implementation details that upgrading will be a pain if they change anything. My ultimate plan here (as it is with the marathon client) is to replace it with our own much simpler, much more testable implementation.

The offending code is here if you're interested: https://github.com/mesosphere/dcos-cli/blob/master/dcos/mesos.py#L31